File: code/2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start

// Definition for singly-linked list.
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* root = new ListNode();
        ListNode* first = root;
        int carry = 0;
        while (l1 != nullptr &&  l2 != nullptr) {
            int val = l1->val + l2->val + carry;
            first->next = new ListNode(val % 10);
            first = first->next;
            carry = val / 10;
            l1 = l1->next;
            l2 = l2->next;
        }
        ListNode * tmp = l1;
        if (l2 != nullptr) {
            tmp = l2;
        }
        while (tmp != nullptr) {
            int val = tmp->val + carry;
            first->next = new ListNode(val % 10);
            first = first->next;
            carry = val / 10;
            if (carry == 0) {
                first->next = tmp->next;
                break;
            }
            tmp = tmp->next;
        }
        if (carry != 0) {
            first->next = new ListNode(carry);
        }
        return root->next;
    }
};
// @lc code=end

File: code/2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        root = ListNode()
        f = root
        carry = 0
        while l1 is not None and l2 is not None: 
            val = l1.val + l2.val + carry
            carry = val // 10
            val = val % 10
            f.next = ListNode(val=val)
            f = f.next
            l1 = l1.next
            l2 = l2.next
        
        l3 = None
        
        if l1 is not None:
            l3 = l1 
        
        if l2 is not None:
            l3 = l2
        
        if carry > 0:
            if l3 is not None:
                while l3 is not None and carry != 0:
                    val = l3.val + carry
                    carry = val // 10
                    val = val % 10
                    f.next = ListNode(val=val)
                    f = f.next
                    l3 = l3.next
                if l3 is None and carry != 0:
                    f.next = ListNode(val=carry)
                else:
                    f.next = l3
            else:
                f.next = ListNode(val=carry)
        else:
            f.next = l3

        return root.next
# @lc code=end

